File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    def rotate(a):
        row, col = len(a), len(a[0])
        result = [[0] * row for _ in range(col)]
        for r in range(row):
            for c in range(col):
                result[c][row - 1 - r] = a[r][c]
        return result

    def check(x, y, b, v):
        if b[x][y] != '0' and b[x][y] == b[x + 1][y] == b[x][y + 1] == b[x + 1][y + 1]:
            v[x][y] = v[x + 1][y] = v[x][y + 1] = v[x + 1][y + 1] = 1
            return True
        else:
            return False

    def delete(b, v):
        c = 0
        for i in range(n):
            for j in range(m - 1, -1, -1):
                if v[i][j] == 1:
                    v[i][j] = 0
                    b[i].pop(j)
                    b[i].append('0')
                    c += 1
        return c

    new_board = rotate([list(b) for b in board])
    visited = [[0] * m for _ in range(n)]

    cnt = answer = 0
    
    while True:
        for a in range(n - 1):
            for b in range(m - 1):
                if check(a, b, new_board, visited):
                    cnt += 1
        
        if cnt <= 0:
            break
        
        answer += delete(new_board, visited)
        cnt = 0

    return answer

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int solution(int m, int n, String[] board) {
        List<List<Character>> newBoard = rotate(board);
        int[][] visited = new int[n][m];
        int answer = 0;

        while (true) {
            int cnt = 0;

            for (int a = 0; a < n - 1; a++) {
                for (int b = 0; b < m - 1; b++) {
                    if (check(a, b, newBoard, visited)) {
                        cnt++;
                    }
                }
            }

            if (cnt <= 0) {
                break;
            }

            answer += delete(m, n, newBoard, visited);
        }

        return answer;
    }

    private List<List<Character>> rotate(String[] board) {
        int row = board.length;
        int col = board[0].length();

        char[][] a = new char[row][col];
        for (int i = 0; i < row; i++) {
            a[i] = board[i].toCharArray();
        }

        List<List<Character>> result = new ArrayList<>();
        for (int c = 0; c < col; c++) {
            List<Character> list = new ArrayList<>();
            for (int r = 0; r < row; r++) {
                list.add(a[row - 1 - r][c]);
            }
            result.add(list);
        }
        return result;
    }

    private boolean check(int x, int y, List<List<Character>> b, int[][] v) {
        char cur = b.get(x).get(y);
        if (cur != '0'
                && cur == b.get(x + 1).get(y)
                && cur == b.get(x).get(y + 1)
                && cur == b.get(x + 1).get(y + 1)) {
            v[x][y] = 1;
            v[x + 1][y] = 1;
            v[x][y + 1] = 1;
            v[x + 1][y + 1] = 1;
            return true;
        }
        return false;
    }

    private int delete(int m, int n, List<List<Character>> b, int[][] v) {
        int c = 0;

        for (int i = 0; i < n; i++) {
            for (int j = m - 1; j >= 0; j--) {
                if (v[i][j] == 1) {
                    v[i][j] = 0;
                    b.get(i).remove(j);
                    b.get(i).add('0');
                    c++;
                }
            }
        }

        return c;
    }
}